File: DSA/binary_trees/tree_includes.py
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

a = Node('a')
b = Node('b')
c = Node('c')
d = Node('d')
e = Node('e')
f = Node('f')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

def tree_includes(root, target):
    if root == None: return []
    queue = [root]

    while queue:
        current = queue.pop(0)
        if current.val == target: return True
        if current.left: queue.append(current.left)
        if current.right: queue.append(current.right)
    return False

def tree_includes_rec(root, target):
    if root == None: return False
    if root.val == target: return True
    return tree_includes_rec(root.left, target) or tree_includes_rec(root.right, target)

print('Searching iteratively')
print(tree_includes(a, 'e'))
print('Searching recursively')
print(tree_includes_rec(a, 'e'))

File: dsa/binary_trees/treeIncludes.js
class Node {
	constructor(val) {
		this.val = val
		this.left = null
		this.right = null
	}
}

const a = new Node('a')
const b = new Node('b')
const c = new Node('c')
const d = new Node('d')
const e = new Node('e')
const f = new Node('f')

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

const treeIncludes = (root, target) => {
	if (root === null) return false
	const queue = [root]

	while (queue.length > 0) {
		const current = queue.shift()
		if (current.val === target) return true
		if (current.left) queue.push(current.left)
		if (current.right) queue.push(current.right)
	}
	return false
}

const treeIncludesRec = (root, target) => {
	if (root === null) return false
	if (root.val === target) return true
	return treeIncludesRec(root.left, target) || treeIncludesRec(root.right, target)
}

console.log('Searching iteratively')
console.log(treeIncludes(a, 'e'))
console.log('Searching recursively')
console.log(treeIncludesRec(a, 'e'))
